From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jiaming Zhang <r772577952@gmail.com>
Date: Wed, 9 Sep 2026 14:59:27 +0800
Subject: [PATCH 2/6] pkg/report, sys, tools/syz-symbolize: miscellaneous
 updates for freebsd

for pkg/report and tools/syz-symbolize, update symbolize function to
normalize line ending in call stack.  Update some syscall descriptions
to avoid errors, and update prepareArch() in sys/syz-extract to extract
some constants.
---
 pkg/report/freebsd.go                 | 53 +++++++++++++++++++++++++++
 sys/freebsd/socket_inet_tcp.txt       |  3 ++
 sys/freebsd/socket_inet_tcp.txt.const |  2 +
 sys/freebsd/sys.txt                   |  2 +
 sys/freebsd/sys.txt.const             |  4 +-
 sys/syz-extract/freebsd.go            | 44 ++++++++++++++++++++++
 tools/syz-symbolize/symbolize.go      |  2 +
 7 files changed, 108 insertions(+), 2 deletions(-)

diff --git a/pkg/report/freebsd.go b/pkg/report/freebsd.go
index 37c50eaf8b..c7512a6e42 100644
--- a/pkg/report/freebsd.go
+++ b/pkg/report/freebsd.go
@@ -4,7 +4,13 @@
 package report
 
 import (
+	"bytes"
+	"fmt"
+	"path/filepath"
 	"regexp"
+	"time"
+
+	"github.com/google/syzkaller/pkg/osutil"
 )
 
 type freebsd struct {
@@ -26,7 +32,54 @@ func (ctx *freebsd) Parse(output []byte) *Report {
 	return simpleLineParser(output, freebsdOopses, freebsdStackParams, ctx.ignores)
 }
 
+// dirty implementation, but at least it works
 func (ctx *freebsd) Symbolize(rep *Report) error {
+	kernelBin := filepath.Join(ctx.kernelDirs.Obj, ctx.target.KernelObject)
+	re := regexp.MustCompile(`#[0-9]+ 0xffffffff`)
+	lines := bytes.Split(rep.Report, []byte("\n"))
+
+	// collect all addresses from the stack trace
+	type pair struct {
+		lnum int
+		addr []byte
+	}
+	var pairs []pair
+	for i, line := range lines {
+		if !re.Match(line) {
+			continue
+		}
+		parts := bytes.Split(line, []byte(" "))
+		if len(parts) < 2 {
+			continue
+		}
+		pairs = append(pairs, pair{lnum: i, addr: parts[1]})
+	}
+
+	// symbolize addresses in batch
+	timeout := 10 * time.Second
+	args := []string{"-e", kernelBin}
+	for _, pair := range pairs {
+		args = append(args, string(pair.addr))
+	}
+	output, err := osutil.RunCmd(timeout, "", "addr2line", args...)
+	if err != nil {
+		return err
+	}
+	outputSlices := bytes.Split(output, []byte("\n"))
+	outputSlices = outputSlices[:len(outputSlices)-1] // remove trailing empty line
+	if len(outputSlices) != len(pairs) {
+		return fmt.Errorf("len(outputSlices)=%v, want %v", len(outputSlices), len(pairs))
+	}
+
+	// append symbolized lines to the end of each line of report
+	for i, pair := range pairs {
+		line := lines[pair.lnum]
+		symbolized := outputSlices[i]
+		symbPath := osutil.Abs(string(symbolized))
+		lines[pair.lnum] = fmt.Appendf(nil, "%s %s", line, symbPath)
+	}
+	rep.Report = bytes.Join(lines, []byte("\r\n"))
+
 	return nil
 }
 
diff --git a/sys/freebsd/socket_inet_tcp.txt b/sys/freebsd/socket_inet_tcp.txt
index 57ede42847..9c6ba7fdfc 100644
--- a/sys/freebsd/socket_inet_tcp.txt
+++ b/sys/freebsd/socket_inet_tcp.txt
@@ -13,6 +13,9 @@ include <netinet/tcp.h>
 include <netinet/cc/cc_newreno.h>
 include <opencrypto/cryptodev.h>
 
+define TCP_PCAP_OUT 2048
+define TCP_PCAP_IN 4096
+
 resource sock_tcp[sock_in]
 
 socket$inet_tcp(domain const[AF_INET], type const[SOCK_STREAM], proto const[0]) sock_tcp
diff --git a/sys/freebsd/socket_inet_tcp.txt.const b/sys/freebsd/socket_inet_tcp.txt.const
index 762b8de25c..5cf0aa7bbc 100644
--- a/sys/freebsd/socket_inet_tcp.txt.const
+++ b/sys/freebsd/socket_inet_tcp.txt.const
@@ -37,6 +37,8 @@ TCP_MD5SIG = 16
 TCP_NODELAY = 1
 TCP_NOOPT = 8
 TCP_NOPUSH = 4
+TCP_PCAP_IN = 4096
+TCP_PCAP_OUT = 2048
 TCP_REMOTE_UDP_ENCAPS_PORT = 71
 TCP_RXTLS_ENABLE = 41
 TCP_RXTLS_MODE = 42
diff --git a/sys/freebsd/sys.txt b/sys/freebsd/sys.txt
index 8738250a38..3f302297d5 100644
--- a/sys/freebsd/sys.txt
+++ b/sys/freebsd/sys.txt
@@ -24,6 +24,8 @@ include <sys/specialfd.h>
 
 type signo int32[0:SIGRTMAX]
 
+define SF_SYNC 4
+
 syz_execute_func(text ptr[in, text[target]]) (disabled)
 
 sf_hdtr {
diff --git a/sys/freebsd/sys.txt.const b/sys/freebsd/sys.txt.const
index 3232654af8..97d0211fdc 100644
--- a/sys/freebsd/sys.txt.const
+++ b/sys/freebsd/sys.txt.const
@@ -147,7 +147,7 @@ SYS_futimesat = 494
 SYS_getegid = 43
 SYS_geteuid = 25
 SYS_getgid = 47
-SYS_getgroups = 79
+SYS_getgroups = 595
 SYS_getitimer = 86
 SYS_getlogin = 49
 SYS_getloginclass = 523
@@ -191,7 +191,7 @@ SYS_setegid = 182
 SYS_seteuid = 183
 SYS_setfib = 175
 SYS_setgid = 181
-SYS_setgroups = 80
+SYS_setgroups = 596
 SYS_setitimer = 83
 SYS_setlogin = 50
 SYS_setloginclass = 524
diff --git a/sys/syz-extract/freebsd.go b/sys/syz-extract/freebsd.go
index 1bcd9823f8..57aa66e011 100644
--- a/sys/syz-extract/freebsd.go
+++ b/sys/syz-extract/freebsd.go
@@ -8,8 +8,10 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
+	"time"
 
 	"github.com/google/syzkaller/pkg/compiler"
+	"github.com/google/syzkaller/pkg/osutil"
 )
 
 type freebsd struct{}
@@ -40,6 +42,46 @@ func (*freebsd) prepareArch(arch *Arch) error {
 		filepath.Join(arch.buildDir, "x86")); err != nil {
 		return fmt.Errorf("failed to create link: %w", err)
 	}
+
+	// Generate necessary headers under sys/syzkaller for const extraction.
+	// device_if.h, bus_if.h, etc..
+	syzDir := filepath.Join(arch.sourceDir, "sys", "syzkaller")
+	awkArgs := []string{
+		"-f",
+		filepath.Join(arch.sourceDir, "sys", "tools", "makeobjops.awk"),
+		filepath.Join(arch.sourceDir, "sys", "kern", "device_if.m"),
+		filepath.Join(arch.sourceDir, "sys", "kern", "bus_if.m"),
+		"-h",
+	}
+	if err := osutil.MkdirAll(syzDir); err != nil {
+		return fmt.Errorf("failed to create directory: %w", err)
+	}
+	if _, err := osutil.RunCmd(10*time.Second, syzDir, "awk", awkArgs...); err != nil {
+		return fmt.Errorf("failed to run awk: %w", err)
+	}
+
+	// Generate vnode_if.h.
+	awkArgs = []string{
+		"-f",
+		filepath.Join(arch.sourceDir, "sys", "tools", "vnode_if.awk"),
+		filepath.Join(arch.sourceDir, "sys", "kern", "vnode_if.src"),
+		"-qhp",
+	}
+	if _, err := osutil.RunCmd(10*time.Second, syzDir, "awk", awkArgs...); err != nil {
+		return fmt.Errorf("failed to run awk: %w", err)
+	}
+
+	// Generate opt_*.h headers and save to sys/syzkaller.
+	confDir := filepath.Join(arch.sourceDir, "sys", "amd64", "conf")
+	configArgs := []string{
+		"-d",
+		syzDir,
+		"GENERIC",
+	}
+	if _, err := osutil.RunCmd(10*time.Second, confDir, "config", configArgs...); err != nil {
+		return fmt.Errorf("failed to run config command: %w", err)
+	}
+
 	return nil
 }
 
@@ -52,10 +94,12 @@ func (*freebsd) processFile(arch *Arch, info *compiler.ConstInfo) (map[string]ui
 		"-D__BSD_VISIBLE=1",
 		"-DCOMPAT_FREEBSD13",
 		"-DCOMPAT_FREEBSD14",
+		"-DCOMPAT_43TTY",
 		"-I", filepath.Join(arch.sourceDir, "sys"),
 		"-I", filepath.Join(arch.sourceDir, "sys", "sys"),
 		"-I", filepath.Join(arch.sourceDir, "sys", "contrib", "ck", "include"),
 		"-I", filepath.Join(arch.sourceDir, "include"),
+		"-I", filepath.Join(arch.sourceDir, "sys", "syzkaller"),
 		"-I", arch.buildDir,
 	}
 	for _, incdir := range info.Incdirs {
diff --git a/tools/syz-symbolize/symbolize.go b/tools/syz-symbolize/symbolize.go
index 0a884119b8..7d01f765b1 100644
--- a/tools/syz-symbolize/symbolize.go
+++ b/tools/syz-symbolize/symbolize.go
@@ -4,6 +4,7 @@
 package main
 
 import (
+	"bytes"
 	"flag"
 	"fmt"
 	"os"
@@ -54,6 +55,7 @@ func main() {
 		tool.Failf("failed to create reporter: %v", err)
 	}
 	text, err := os.ReadFile(flag.Args()[0])
+	text = bytes.Replace(text, []byte("\r\n"), []byte("\n"), -1)
 	if err != nil {
 		tool.Failf("failed to open input file: %v", err)
 	}
